include/deal.II/base/array_view.h

```diff
@@ -0,0 +1,219 @@
+
+#ifndef dealii__array_view_h
+#define dealii__array_view_h
+
+#include <deal.II/base/config.h>
+#include <deal.II/base/exceptions.h>
+#include <boost/type_traits/remove_cv.hpp>
+#include <boost/type_traits/is_same.hpp>
+#include <boost/utility/enable_if.hpp>
+
+#include <vector>
+
+DEAL_II_NAMESPACE_OPEN
+
+
+/**
+ * A class that represents a window of memory locations of type
+ * @p ElementType and presents it as if it was an array that
+ * can be accessed via an <code>operator[]</code>. In essence,
+ * this class is nothing more than just a pointer to the first
+ * location and an integer that represents the length of the array
+ * in elements.
+ *
+ * The advantage of using this class is that you don't have to pass
+ * around pairs of pointers and that <code>operator[]</code> checks
+ * for the validity of the index with which you subscript this
+ * array view.
+ *
+ * This class can handle views to both non-constant and constant memory
+ * locations. If you want to represent a view of a constant array,
+ * then the template argument type of this class needs to be
+ * @p const as well. The following code snippet gives an example:
+ * @code
+ *   std::vector<int>       array       = get_data();  // a writable array
+ *
+ *   ArrayView<int> view (&array[5], 5);               // a view of elements 5..9 (inclusive)
+ *   view[2] = 42;                                     // array[7] is set to 42
+ *
+ *   ArrayView<const int> const_view (&array[5], 5);   // same view, but read-only
+ *   int element_7 = const_view[2];                    // returns 42
+ *   const_view[2] = 42;                               // error, can't write into this view
+ * @endcode
+ *
+ * @ingroup data
+ * @author Wolfgang Bangerth, 2015
+ */
+template <typename ElementType>
+class ArrayView
+{
+public:
+  /**
+   * A typedef that denotes the "value_type" of this container-like
+   * class, i.e., the type of the element it "stores" or points to.
+   */
+  typedef ElementType value_type;
+
+  /**
+   * Constructor.
+   * @param[in] starting_element A pointer to the first element of the array
+   *   this object should represent.
+   * @param[in] n_elements The length (in elements) of the chunk of memory
+   *   this object should represent.
+   */
+  ArrayView (value_type       *starting_element,
+             const std::size_t n_elements);
+
+  /**
+   * Copy constructor from array views that point to non-@p const elements. If
+   * the current object will point to non-@p const elements, then this is a
+   * straight forward copy constructor. On the other hand, if the current type's
+   * @p ElementType template argument is a @p const qualified type, then the
+   * current constructor is a conversion constructor that converts a
+   * non-@p const view to a @p const view, akin to converting a non-@p const
+   * pointer to a @p const pointer.
+   */
+  ArrayView (const ArrayView<typename boost::remove_cv<value_type>::type> &view);
```

Comment:
I think it would be safer to make these explicit or are they implicit on purpose?